In [1091]:
import pandas as pd
import numpy as np
import os
import warnings
warnings.filterwarnings(action = 'ignore')

In [1092]:
DATA_DIR = './csv_files'
data_files = os.listdir(DATA_DIR)[1:]
data_files

['test.csv', 'train.csv']

In [1093]:
data_files = reversed([os.path.join(DATA_DIR, filename) for filename in os.listdir(DATA_DIR)[1:]])

df_list = []

for filename in data_files:
    df_list.append(pd.read_csv(filename))

### Test Data

In [1094]:
df_list[0]

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0        1          60       RL         65.0     8450   Pave   NaN      Reg   
1        2          20       RL         80.0     9600   Pave   NaN      Reg   
2        3          60       RL         68.0    11250   Pave   NaN      IR1   
3        4          70       RL         60.0     9550   Pave   NaN      IR1   
4        5          60       RL         84.0    14260   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1455  1456          60       RL         62.0     7917   Pave   NaN      Reg   
1456  1457          20       RL         85.0    13175   Pave   NaN      Reg   
1457  1458          70       RL         66.0     9042   Pave   NaN      Reg   
1458  1459          20       RL         68.0     9717   Pave   NaN      Reg   
1459  1460          20       RL         75.0     9937   Pave   NaN      Reg   

     LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
2            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
3            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
4            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
...          ...       ...  ...      ...    ...    ...         ...     ...   
1455         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1456         Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
1457         Lvl    AllPub  ...        0    NaN  GdPrv        Shed    2500   
1458         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1459         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice  
0         2   2008        WD         Normal     208500  
1         5   2007        WD         Normal     181500  
2         9   2008        WD         Normal     223500  
3         2   2006        WD        Abnorml     140000  
4        12   2008        WD         Normal     250000  
...     ...    ...       ...            ...        ...  
1455      8   2007        WD         Normal     175000  
1456      2   2010        WD         Normal     210000  
1457      5   2010        WD         Normal     266500  
1458      4   2010        WD         Normal     142125  
1459      6   2008        WD         Normal     147500  

[1460 rows x 81 columns]

### Train Data

In [1095]:
df_list[1]

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0     1461          20       RH         80.0    11622   Pave   NaN      Reg   
1     1462          20       RL         81.0    14267   Pave   NaN      IR1   
2     1463          60       RL         74.0    13830   Pave   NaN      IR1   
3     1464          60       RL         78.0     9978   Pave   NaN      IR1   
4     1465         120       RL         43.0     5005   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1454  2915         160       RM         21.0     1936   Pave   NaN      Reg   
1455  2916         160       RM         21.0     1894   Pave   NaN      Reg   
1456  2917          20       RL        160.0    20000   Pave   NaN      Reg   
1457  2918          85       RL         62.0    10441   Pave   NaN      Reg   
1458  2919          60       RL         74.0     9627   Pave   NaN      Reg   

     LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence  \
0            Lvl    AllPub  ...         120        0    NaN  MnPrv   
1            Lvl    AllPub  ...           0        0    NaN    NaN   
2            Lvl    AllPub  ...           0        0    NaN  MnPrv   
3            Lvl    AllPub  ...           0        0    NaN    NaN   
4            HLS    AllPub  ...         144        0    NaN    NaN   
...          ...       ...  ...         ...      ...    ...    ...   
1454         Lvl    AllPub  ...           0        0    NaN    NaN   
1455         Lvl    AllPub  ...           0        0    NaN    NaN   
1456         Lvl    AllPub  ...           0        0    NaN    NaN   
1457         Lvl    AllPub  ...           0        0    NaN  MnPrv   
1458         Lvl    AllPub  ...           0        0    NaN    NaN   

     MiscFeature MiscVal MoSold  YrSold  SaleType  SaleCondition  
0            NaN       0      6    2010        WD         Normal  
1           Gar2   12500      6    2010        WD         Normal  
2            NaN       0      3    2010        WD         Normal  
3            NaN       0      6    2010        WD         Normal  
4            NaN       0      1    2010        WD         Normal  
...          ...     ...    ...     ...       ...            ...  
1454         NaN       0      6    2006        WD         Normal  
1455         NaN       0      4    2006        WD        Abnorml  
1456         NaN       0      9    2006        WD        Abnorml  
1457        Shed     700      7    2006        WD         Normal  
1458         NaN       0     11    2006        WD         Normal  

[1459 rows x 80 columns]

### 모판 만들기

In [1096]:
df = pd.concat(df_list, sort= False).set_index('Id')
df

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                      
1             60       RL         65.0     8450   Pave   NaN      Reg   
2             20       RL         80.0     9600   Pave   NaN      Reg   
3             60       RL         68.0    11250   Pave   NaN      IR1   
4             70       RL         60.0     9550   Pave   NaN      IR1   
5             60       RL         84.0    14260   Pave   NaN      IR1   
...          ...      ...          ...      ...    ...   ...      ...   
2915         160       RM         21.0     1936   Pave   NaN      Reg   
2916         160       RM         21.0     1894   Pave   NaN      Reg   
2917          20       RL        160.0    20000   Pave   NaN      Reg   
2918          85       RL         62.0    10441   Pave   NaN      Reg   
2919          60       RL         74.0     9627   Pave   NaN      Reg   

     LandContour Utilities LotConfig  ... PoolArea PoolQC  Fence MiscFeature  \
Id                                    ...                                      
1            Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
2            Lvl    AllPub       FR2  ...        0    NaN    NaN         NaN   
3            Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
4            Lvl    AllPub    Corner  ...        0    NaN    NaN         NaN   
5            Lvl    AllPub       FR2  ...        0    NaN    NaN         NaN   
...          ...       ...       ...  ...      ...    ...    ...         ...   
2915         Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
2916         Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
2917         Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   
2918         Lvl    AllPub    Inside  ...        0    NaN  MnPrv        Shed   
2919         Lvl    AllPub    Inside  ...        0    NaN    NaN         NaN   

     MiscVal MoSold  YrSold  SaleType  SaleCondition  SalePrice  
Id                                                               
1          0      2    2008        WD         Normal   208500.0  
2          0      5    2007        WD         Normal   181500.0  
3          0      9    2008        WD         Normal   223500.0  
4          0      2    2006        WD        Abnorml   140000.0  
5          0     12    2008        WD         Normal   250000.0  
...      ...    ...     ...       ...            ...        ...  
2915       0      6    2006        WD         Normal        NaN  
2916       0      4    2006        WD        Abnorml        NaN  
2917       0      9    2006        WD        Abnorml        NaN  
2918     700      7    2006        WD         Normal        NaN  
2919       0     11    2006        WD         Normal        NaN  

[2919 rows x 80 columns]

추후에 Train, Test Data 구분하기 위해서 Train, Test Data index 파악

In [1097]:
number_of_train_df = df.SalePrice.notnull().sum()
number_of_test_df = df.SalePrice.isnull().sum()
number_of_train_df, number_of_test_df

(1460, 1459)

예측할 Data Column : SalePrice

In [1098]:
y_true = df.pop('SalePrice')[:number_of_train_df]
y_true

Id
1       208500.0
2       181500.0
3       223500.0
4       140000.0
5       250000.0
          ...   
1456    175000.0
1457    210000.0
1458    266500.0
1459    142125.0
1460    147500.0
Name: SalePrice, Length: 1460, dtype: float64

### 결측치 확인

In [1099]:
df_null_check = df.isnull().sum()
df_isnull = df_null_check[df_null_check > 0]
df_isnull

MSZoning           4
LotFrontage      486
Alley           2721
Utilities          2
Exterior1st        1
Exterior2nd        1
MasVnrType        24
MasVnrArea        23
BsmtQual          81
BsmtCond          82
BsmtExposure      82
BsmtFinType1      79
BsmtFinSF1         1
BsmtFinType2      80
BsmtFinSF2         1
BsmtUnfSF          1
TotalBsmtSF        1
Electrical         1
BsmtFullBath       2
BsmtHalfBath       2
KitchenQual        1
Functional         2
FireplaceQu     1420
GarageType       157
GarageYrBlt      159
GarageFinish     159
GarageCars         1
GarageArea         1
GarageQual       159
GarageCond       159
PoolQC          2909
Fence           2348
MiscFeature     2814
SaleType           1
dtype: int64

### 결측치 비율 보기

In [1100]:
df_isnull_percent = df_isnull / len(df) * 100
df_isnull_over_50 = df_isnull_percent[df_isnull_percent > 50]
df_isnull_over_50

Alley          93.216855
PoolQC         99.657417
Fence          80.438506
MiscFeature    96.402878
dtype: float64

결측치가 50%이상인 Column 삭제

In [1101]:
df = df.drop(df_isnull_over_50.index, axis = 1)
df

MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
Id                                                                            
1             60       RL         65.0     8450   Pave      Reg         Lvl   
2             20       RL         80.0     9600   Pave      Reg         Lvl   
3             60       RL         68.0    11250   Pave      IR1         Lvl   
4             70       RL         60.0     9550   Pave      IR1         Lvl   
5             60       RL         84.0    14260   Pave      IR1         Lvl   
...          ...      ...          ...      ...    ...      ...         ...   
2915         160       RM         21.0     1936   Pave      Reg         Lvl   
2916         160       RM         21.0     1894   Pave      Reg         Lvl   
2917          20       RL        160.0    20000   Pave      Reg         Lvl   
2918          85       RL         62.0    10441   Pave      Reg         Lvl   
2919          60       RL         74.0     9627   Pave      Reg         Lvl   

     Utilities LotConfig LandSlope  ... OpenPorchSF EnclosedPorch 3SsnPorch  \
Id                                  ...                                       
1       AllPub    Inside       Gtl  ...          61             0         0   
2       AllPub       FR2       Gtl  ...           0             0         0   
3       AllPub    Inside       Gtl  ...          42             0         0   
4       AllPub    Corner       Gtl  ...          35           272         0   
5       AllPub       FR2       Gtl  ...          84             0         0   
...        ...       ...       ...  ...         ...           ...       ...   
2915    AllPub    Inside       Gtl  ...           0             0         0   
2916    AllPub    Inside       Gtl  ...          24             0         0   
2917    AllPub    Inside       Gtl  ...           0             0         0   
2918    AllPub    Inside       Gtl  ...          32             0         0   
2919    AllPub    Inside       Mod  ...          48             0         0   

     ScreenPorch PoolArea  MiscVal  MoSold  YrSold  SaleType SaleCondition  
Id                                                                          
1              0        0        0       2    2008        WD        Normal  
2              0        0        0       5    2007        WD        Normal  
3              0        0        0       9    2008        WD        Normal  
4              0        0        0       2    2006        WD       Abnorml  
5              0        0        0      12    2008        WD        Normal  
...          ...      ...      ...     ...     ...       ...           ...  
2915           0        0        0       6    2006        WD        Normal  
2916           0        0        0       4    2006        WD       Abnorml  
2917           0        0        0       9    2006        WD       Abnorml  
2918           0        0      700       7    2006        WD        Normal  
2919           0        0        0      11    2006        WD        Normal  

[2919 rows x 75 columns]

### Data 확인

In [1102]:
df_dtypes = df.dtypes
df_dtypes

MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
Street            object
                  ...   
MiscVal            int64
MoSold             int64
YrSold             int64
SaleType          object
SaleCondition     object
Length: 75, dtype: object

In [1443]:
# df_numeric_columns = np.array(df_dtypes[(df_dtypes == 'int64') | (df_dtypes == float)].index)
# df_numeric_columns
# df_object_columns = np.array(df_dtypes[df_dtypes == object].index)
# df_object_columns

In [1444]:
numeric_columns= df.select_dtypes(["int", "float"]).columns
object_columns = df.select_dtypes("object").columns

### 결측치 처리

In [1105]:
df_isnull = df.isnull().sum()
df_isnull_data = df_isnull[df_isnull>0]
df_isnull_columns = np.array(df_isnull_data.index)
df_isnull_columns

array(['MSZoning', 'LotFrontage', 'Utilities', 'Exterior1st',
       'Exterior2nd', 'MasVnrType', 'MasVnrArea', 'BsmtQual', 'BsmtCond',
       'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1', 'BsmtFinType2',
       'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Electrical',
       'BsmtFullBath', 'BsmtHalfBath', 'KitchenQual', 'Functional',
       'FireplaceQu', 'GarageType', 'GarageYrBlt', 'GarageFinish',
       'GarageCars', 'GarageArea', 'GarageQual', 'GarageCond', 'SaleType'],
      dtype=object)

### MSZoning 일반 구역 분류
: Identifies the general zoning classification of the sale.
		
        A	: Agriculture 농업
        C	: Commercial  상업
        FV : Floating Village Residential 부유마을주택
        I	: Industrial 산업, 공업
        RH : Residential High Density 고밀도 거주
        RL : Residential Low Density 저밀도 거주
        RP : Residential Low Density Park 저밀도 거주 공원
        RM : Residential Medium Density 중간밀도 거주

#### MSZoing의 결측치는 MSSubClass와 비교하여 결측치 처리

- MSSubClass 주거유형  
: Identifies the type of dwelling involved in the sale.

        20	1-STORY 1946 & NEWER ALL STYLES
        30	1-STORY 1945 & OLDER
        40	1-STORY W/FINISHED ATTIC ALL AGES
        45	1-1/2 STORY - UNFINISHED ALL AGES
        50	1-1/2 STORY FINISHED ALL AGES
        60	2-STORY 1946 & NEWER
        70	2-STORY 1945 & OLDER
        75	2-1/2 STORY ALL AGES
        80	SPLIT OR MULTI-LEVEL
        85	SPLIT FOYER
        90	DUPLEX - ALL STYLES AND AGES
       120	1-STORY PUD (Planned Unit Development) - 1946 & NEWER
       150	1-1/2 STORY PUD - ALL AGES
       160	2-STORY PUD - 1946 & NEWER
       180	PUD - MULTILEVEL - INCL SPLIT LEV/FOYER
       190	2 FAMILY CONVERSION - ALL STYLES AND AGES

In [1106]:
df['MSZoning'].value_counts()

RL         2265
RM          460
FV          139
RH           26
C (all)      25
Name: MSZoning, dtype: int64

MSSubClass 결측치 확인

In [1107]:
df['MSSubClass'].isnull().sum()

0

In [1108]:
df[df['MSZoning'].isnull()][['MSZoning', 'MSSubClass']]

MSZoning  MSSubClass
Id                       
1916      NaN          30
2217      NaN          20
2251      NaN          70
2905      NaN          20

In [1109]:
df.groupby('MSSubClass')['MSZoning'].describe()

count unique top  freq
MSSubClass                       
20          1077      5  RL  1016
30           138      4  RM    67
40             6      2  RL     4
45            18      3  RM    11
50           287      4  RL   159
60           575      3  RL   529
70           127      4  RM    63
75            23      2  RM    14
80           118      2  RL   115
85            48      2  RL    47
90           109      3  RL    92
120          182      4  RL   117
150            1      1  RL     1
160          128      3  RM    64
180           17      1  RM    17
190           61      4  RL    31

In [1110]:
df['MSZoning'] = df.groupby('MSSubClass')['MSZoning'].apply(lambda x : x.fillna(x.mode().iloc[0]))

In [1111]:
df['MSZoning'].isnull().sum()

0

In [1112]:
df.loc[2905][['MSZoning', 'MSSubClass']]

MSZoning      RL
MSSubClass    20
Name: 2905, dtype: object

### LotFrontage 부동산에 연결된 거리의 선형 피트
: Linear feet of street connected to property

In [1113]:
df['LotFrontage'].isnull().sum()

486

In [1114]:
df['LotArea'].isnull().sum()

0

In [1115]:
def minmax(input_data):
    return input_data.apply(lambda x : (x-np.min(x))/(np.max(x)-np.min(x)))

In [1116]:
LotFrontage_LotArea_minmax = minmax(df[['LotArea', df_isnull_columns[1]]])
LotFrontage_LotArea_minmax.head()

LotArea  LotFrontage
Id                       
1   0.033420     0.150685
2   0.038795     0.202055
3   0.046507     0.160959
4   0.038561     0.133562
5   0.060576     0.215753

In [1117]:
corr = LotFrontage_LotArea_minmax.corr()
corr

LotArea  LotFrontage
LotArea      1.000000     0.489896
LotFrontage  0.489896     1.000000

In [1118]:
LotFrontage_LotArea_isnull =df[df['LotFrontage'].isnull()][['LotFrontage', 'LotArea']]
LotFrontage_LotArea_isnull.head()

LotFrontage  LotArea
Id                      
8           NaN    10382
13          NaN    12968
15          NaN    10920
17          NaN    11241
25          NaN     8246

In [1119]:
LotFrontage_LotArea_notnull = df[df['LotFrontage'].notnull()][['LotFrontage','LotArea']]
LotFrontage_LotArea_notnull.head()

LotFrontage  LotArea
Id                      
1          65.0     8450
2          80.0     9600
3          68.0    11250
4          60.0     9550
5          84.0    14260

In [1120]:
LotFrontage_LotArea_notnull.groupby('LotArea').describe()

LotFrontage                                               
              count   mean  std    min    25%    50%    75%    max
LotArea                                                           
1300            1.0   24.0  NaN   24.0   24.0   24.0   24.0   24.0
1470            1.0   21.0  NaN   21.0   21.0   21.0   21.0   21.0
1476            1.0   21.0  NaN   21.0   21.0   21.0   21.0   21.0
1477            2.0   21.0  0.0   21.0   21.0   21.0   21.0   21.0
1484            1.0   21.0  NaN   21.0   21.0   21.0   21.0   21.0
...             ...    ...  ...    ...    ...    ...    ...    ...
51974           1.0   52.0  NaN   52.0   52.0   52.0   52.0   52.0
53504           1.0   47.0  NaN   47.0   47.0   47.0   47.0   47.0
63887           1.0  313.0  NaN  313.0  313.0  313.0  313.0  313.0
70761           1.0   62.0  NaN   62.0   62.0   62.0   62.0   62.0
215245          1.0  150.0  NaN  150.0  150.0  150.0  150.0  150.0

[1598 rows x 8 columns]

- Transform 사용하여 결측치 처리

In [1121]:
df['LotFrontage'].fillna(df.groupby('LotArea')['LotFrontage'].transform('mean'), inplace=True)

- 아직 결측치가 완전히 처리 되지 않았다. 

In [1122]:
df['LotFrontage'].isnull().sum()

372

- 나머지 결측치는 평균으로 채우겠다.

In [1123]:
df['LotFrontage'].fillna(df['LotFrontage'].mean(), inplace = True)

- 결측치 처리 확인

In [1124]:
df['LotFrontage'].isnull().sum()

0

### Utilities
: Type of utilities available
		
- AllPub : All public Utilities (E,G,W,& S)	모든 공공 사업  
- NoSewr : Electricity, Gas, and Water (Septic Tank) 전기, 가스, 수도(세정 탱크)  
- NoSeWa : Electricity and Gas Only 전기, 가스 전용  
- ELO : Electricity only	전기만  

In [1125]:
df['Utilities'].isnull().sum()

2

In [1126]:
LandContour_Utilities_notnull = df[df[['LandContour','Utilities']]['Utilities'].notnull()]
LandContour_Utilities_notnull.groupby('LandContour')['Utilities'].describe()

count unique     top  freq
LandContour                           
Bnk           116      1  AllPub   116
HLS           120      1  AllPub   120
Low            60      1  AllPub    60
Lvl          2621      2  AllPub  2620

- Utilities 결측치를 AllPub으로 채우기

In [1127]:
Utilities_mode = df['Utilities'].mode().iloc[0]
Utilities_mode

'AllPub'

In [1128]:
df['Utilities'][df['Utilities'].isnull()] = Utilities_mode

In [1129]:
df['Utilities'].isnull().sum()

0

### Exterior1st
: Exterior covering on house

       AsbShng	Asbestos Shingles
       AsphShn	Asphalt Shingles
       BrkComm	Brick Common
       BrkFace	Brick Face
       CBlock	Cinder Block
       CemntBd	Cement Board
       HdBoard	Hard Board
       ImStucc	Imitation Stucco
       MetalSd	Metal Siding
       Other	Other
       Plywood	Plywood
       PreCast	PreCast	
       Stone	Stone
       Stucco	Stucco
       VinylSd	Vinyl Siding
       Wd Sdng	Wood Siding
       WdShing	Wood Shingles

In [1130]:
df['Exterior1st'].value_counts()

VinylSd    1025
MetalSd     450
HdBoard     442
Wd Sdng     411
Plywood     221
CemntBd     126
BrkFace      87
WdShing      56
AsbShng      44
Stucco       43
BrkComm       6
AsphShn       2
Stone         2
CBlock        2
ImStucc       1
Name: Exterior1st, dtype: int64

In [1131]:
df['Exterior1st'].isnull().sum()

1

In [1132]:
Exterior1st_Exterior2nd = df[['Exterior1st', 'Exterior2nd']]
Exterior1st_Exterior2nd[Exterior1st_Exterior2nd['Exterior1st'].isnull()]

Exterior1st Exterior2nd
Id                          
2152         NaN         NaN

In [1133]:
df['ExterCond'].isnull().sum()

0

In [1134]:
Exterior1st_Foundation = df[['Exterior1st', 'Foundation']]
Exterior1st_Foundation[Exterior1st_Foundation['Exterior1st'].isnull()]

Exterior1st Foundation
Id                         
2152         NaN      PConc

- Exterior1st 결측치 Foundation에 따라 채우기

In [1135]:
Exterior1st_Foundation.groupby('Foundation')['Exterior1st'].describe()

count unique      top freq
Foundation                           
BrkTil       311     10  Wd Sdng  126
CBlock      1235     14  HdBoard  313
PConc       1307     14  VinylSd  845
Slab          49     10  HdBoard   10
Stone         11      6  Wd Sdng    4
Wood           5      3  VinylSd    3

In [1136]:
df['Exterior1st'] = df.groupby('Foundation')['Exterior1st'].apply(lambda x: x.fillna(x.mode().iloc[0]))

In [1137]:
df['Exterior1st'].isnull().sum()

0

In [1138]:
df.iloc[2151][['Exterior1st', 'Foundation']]

Exterior1st    VinylSd
Foundation       PConc
Name: 2152, dtype: object

### Exterior2nd
: Exterior covering on house (if more than one material)

       AsbShng	Asbestos Shingles
       AsphShn	Asphalt Shingles
       BrkComm	Brick Common
       BrkFace	Brick Face
       CBlock	Cinder Block
       CemntBd	Cement Board
       HdBoard	Hard Board
       ImStucc	Imitation Stucco
       MetalSd	Metal Siding
       Other	Other
       Plywood	Plywood
       PreCast	PreCast
       Stone	Stone
       Stucco	Stucco
       VinylSd	Vinyl Siding
       Wd Sdng	Wood Siding
       WdShing	Wood Shingles

In [1139]:
df['Exterior2nd'].isnull().sum()

1

In [1140]:
Exterior1st_Exterior2nd.groupby('Exterior1st')['Exterior2nd'].describe()

count unique      top  freq
Exterior1st                            
AsbShng        44      6  AsbShng    35
AsphShn         2      1  AsphShn     2
BrkComm         6      3  Brk Cmn     4
BrkFace        87     10  BrkFace    44
CBlock          2      2   CBlock     1
CemntBd       126      3  CmentBd   124
HdBoard       442      9  HdBoard   383
ImStucc         1      1  ImStucc     1
MetalSd       450      8  MetalSd   437
Plywood       221      8  Plywood   186
Stone           2      2  HdBoard     1
Stucco         43      7   Stucco    33
VinylSd      1025     10  VinylSd  1005
Wd Sdng       411     11  Wd Sdng   353
WdShing        56      5  Wd Shng    41

- Exterior2nd 결측치는 Exterior1st에 따라 처리

In [1141]:
df['Exterior2nd'] = df.groupby('Foundation')['Exterior2nd'].apply(lambda x: x.fillna(x.mode().iloc[0]))

In [1142]:
df['Exterior2nd'].isnull().sum()

0

In [1143]:
df.iloc[2151][['Exterior1st','Exterior2nd']]

Exterior1st    VinylSd
Exterior2nd    VinylSd
Name: 2152, dtype: object

### MasVnrType, MasVnrArea
MasVnrType : Masonry veneer type

       BrkCmn	Brick Common
       BrkFace	Brick Face
       CBlock	Cinder Block
       None	        None
       Stone	Stone

MasVnrArea : Masonry veneer area in square feet

In [1144]:
df['MasVnrType'].value_counts()

None       1742
BrkFace     879
Stone       249
BrkCmn       25
Name: MasVnrType, dtype: int64

In [1145]:
df['MasVnrType'].isnull().sum()

24

In [1146]:
df['MasVnrArea'].isnull().sum()

23

In [1147]:
df[df['MasVnrType'].isnull()][['MasVnrType', 'MasVnrArea']]

MasVnrType  MasVnrArea
Id                         
235         NaN         NaN
530         NaN         NaN
651         NaN         NaN
937         NaN         NaN
974         NaN         NaN
978         NaN         NaN
1244        NaN         NaN
1279        NaN         NaN
1692        NaN         NaN
1707        NaN         NaN
1883        NaN         NaN
1993        NaN         NaN
2005        NaN         NaN
2042        NaN         NaN
2312        NaN         NaN
2326        NaN         NaN
2341        NaN         NaN
2350        NaN         NaN
2369        NaN         NaN
2593        NaN         NaN
2611        NaN       198.0
2658        NaN         NaN
2687        NaN         NaN
2863        NaN         NaN

- MasVnrArea부터 채운 후 MasVnrType 결측치 처리

이전에 데이터 상관계수 도출했을 때 OverallQual이 0.407의 상관을 가졌음

In [1148]:
df['OverallQual'].isnull().sum()

0

In [1149]:
df.groupby('OverallQual')['MasVnrArea'].describe()

count        mean         std  min    25%    50%     75%     max
OverallQual                                                                  
1              4.0    0.000000    0.000000  0.0    0.0    0.0    0.00     0.0
2             13.0    0.000000    0.000000  0.0    0.0    0.0    0.00     0.0
3             40.0    8.500000   53.758720  0.0    0.0    0.0    0.00   340.0
4            226.0   15.783186   69.371668  0.0    0.0    0.0    0.00   771.0
5            825.0   43.101818  103.264491  0.0    0.0    0.0    0.00  1047.0
6            727.0  100.587345  163.040270  0.0    0.0    0.0  170.00  1224.0
7            590.0   96.849153  156.864424  0.0    0.0    0.0  170.00  1600.0
8            334.0  205.631737  217.679529  0.0    0.0  177.0  309.75  1290.0
9            107.0  381.373832  236.153633  0.0  215.0  362.0  521.50  1095.0
10            30.0  558.433333  388.063689  0.0  244.0  567.0  787.50  1378.0

- MasVnrArea 결측치 처리

In [1150]:
df.groupby('OverallQual')['MasVnrArea'].mean()

OverallQual
1       0.000000
2       0.000000
3       8.500000
4      15.783186
5      43.101818
6     100.587345
7      96.849153
8     205.631737
9     381.373832
10    558.433333
Name: MasVnrArea, dtype: float64

In [1151]:
df[['OverallQual','MasVnrArea']][df['MasVnrArea'].isnull()].head()

OverallQual  MasVnrArea
Id                          
235            6         NaN
530            6         NaN
651            7         NaN
937            7         NaN
974            7         NaN

In [1152]:
MasVnrArea_mean_round = round(df.groupby('OverallQual')['MasVnrArea'].transform('mean'), 0)

In [1153]:
df['MasVnrArea'].fillna(MasVnrArea_mean_round, inplace=True)

In [1154]:
df['MasVnrArea'].iloc[[2311, 2862]]

Id
2312    101.0
2863    101.0
Name: MasVnrArea, dtype: float64

- MasVnrArea 결측치 처리 확인

In [1155]:
df['MasVnrArea'].isnull().sum()

0

- MasVnrType 결측치 처리

In [1156]:
MasVnrType_MasVnrArea = df[['MasVnrType', 'MasVnrArea']]
MasVnrType_MasVnrArea.head()

MasVnrType  MasVnrArea
Id                       
1     BrkFace       196.0
2        None         0.0
3     BrkFace       162.0
4        None         0.0
5     BrkFace       350.0

In [1157]:
MasVnrType_MasVnrArea[MasVnrType_MasVnrArea['MasVnrType'].isnull()].head()

MasVnrType  MasVnrArea
Id                        
235        NaN       101.0
530        NaN       101.0
651        NaN        97.0
937        NaN        97.0
974        NaN        97.0

In [1158]:
df.groupby('MasVnrArea')['MasVnrType'].describe()

count unique      top  freq
MasVnrArea                            
0.0         1738      3     None  1735
1.0            3      1     None     3
3.0            1      1  BrkFace     1
11.0           1      1  BrkFace     1
14.0           4      1  BrkFace     4
...          ...    ...      ...   ...
1170.0         1      1  BrkFace     1
1224.0         2      2    Stone     1
1290.0         1      1  BrkFace     1
1378.0         1      1  BrkFace     1
1600.0         1      1  BrkFace     1

[445 rows x 4 columns]

In [1159]:
MasVnrType_top = df.groupby('MasVnrArea')['MasVnrType'].describe().top

In [1160]:
df['MasVnrType'].fillna(MasVnrType_top, inplace=True)

In [1161]:
df['MasVnrType'].isnull().sum()

21

In [1162]:
df.groupby(['OverallQual','MasVnrArea'])['MasVnrType'].describe()

count unique      top freq
OverallQual MasVnrArea                           
1           0.0            4      1     None    4
2           0.0           13      1     None   13
3           0.0           39      1     None   39
            340.0          1      1  BrkFace    1
4           0.0          203      1     None  203
...                      ...    ...      ...  ...
10          970.0          1      1  BrkFace    1
            1031.0         1      1  BrkFace    1
            1170.0         1      1  BrkFace    1
            1224.0         1      1    Stone    1
            1378.0         1      1  BrkFace    1

[811 rows x 4 columns]

- MasVnrType의 나머지 결측치는 BrkFace로 채우기

In [1163]:
MVtype_mode = df[df['MasVnrArea'] != 0][df_isnull_columns[5]].mode().iloc[0]
MVtype_mode

'BrkFace'

In [1164]:
df['MasVnrType'].fillna(MVtype_mode, inplace=True)

In [1165]:
df['MasVnrType'].isnull().sum()

0

### BsmtQual
: Evaluates the height of the basement

       Ex	Excellent (100+ inches)	
       Gd	Good (90-99 inches)
       TA	Typical (80-89 inches)
       Fa	Fair (70-79 inches)
       Po	Poor (<70 inches
       NA	No Basement

In [1166]:
df['BsmtQual'].value_counts()

TA    1283
Gd    1209
Ex     258
Fa      88
Name: BsmtQual, dtype: int64

In [1167]:
df['BsmtQual'].isnull().sum()

81

In [1168]:
Bsmt = []
for i in range(len(df_isnull_columns)):
    if 'Bsmt' in df_isnull_columns[i]:
        Bsmt.append(df_isnull_columns[i])
    else:
        pass
Bsmt

['BsmtQual',
 'BsmtCond',
 'BsmtExposure',
 'BsmtFinType1',
 'BsmtFinSF1',
 'BsmtFinType2',
 'BsmtFinSF2',
 'BsmtUnfSF',
 'TotalBsmtSF',
 'BsmtFullBath',
 'BsmtHalfBath']

In [1169]:
df[Bsmt].isnull().sum()

BsmtQual        81
BsmtCond        82
BsmtExposure    82
BsmtFinType1    79
BsmtFinSF1       1
BsmtFinType2    80
BsmtFinSF2       1
BsmtUnfSF        1
TotalBsmtSF      1
BsmtFullBath     2
BsmtHalfBath     2
dtype: int64

In [1170]:
Bsmt_df = df[Bsmt]
Bsmt_df[df['BsmtQual'].isnull()]

BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
Id                                                                          
18        NaN      NaN          NaN          NaN         0.0          NaN   
40        NaN      NaN          NaN          NaN         0.0          NaN   
91        NaN      NaN          NaN          NaN         0.0          NaN   
103       NaN      NaN          NaN          NaN         0.0          NaN   
157       NaN      NaN          NaN          NaN         0.0          NaN   
...       ...      ...          ...          ...         ...          ...   
2804      NaN      NaN          NaN          NaN         0.0          NaN   
2805      NaN      NaN          NaN          NaN         0.0          NaN   
2825      NaN      NaN          NaN          NaN         0.0          NaN   
2892      NaN      NaN          NaN          NaN         0.0          NaN   
2905      NaN      NaN          NaN          NaN         0.0          NaN   

      BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  BsmtFullBath  BsmtHalfBath  
Id                                                                    
18           0.0        0.0          0.0           0.0           0.0  
40           0.0        0.0          0.0           0.0           0.0  
91           0.0        0.0          0.0           0.0           0.0  
103          0.0        0.0          0.0           0.0           0.0  
157          0.0        0.0          0.0           0.0           0.0  
...          ...        ...          ...           ...           ...  
2804         0.0        0.0          0.0           0.0           0.0  
2805         0.0        0.0          0.0           0.0           0.0  
2825         0.0        0.0          0.0           0.0           0.0  
2892         0.0        0.0          0.0           0.0           0.0  
2905         0.0        0.0          0.0           0.0           0.0  

[81 rows x 11 columns]

- 행이 2121인 결측치들은 최빈값으로 처리

In [1171]:
Bsmt_mode = df[Bsmt].mode()
Bsmt_mode

BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
0       TA       TA           No          Unf         0.0          Unf   

   BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  BsmtFullBath  BsmtHalfBath  
0         0.0        0.0          0.0           0.0           0.0

In [1172]:
df.loc[2121,Bsmt] = Bsmt_mode.iloc[0]

In [1173]:
df.loc[2121,Bsmt]

BsmtQual         TA
BsmtCond         TA
BsmtExposure     No
BsmtFinType1    Unf
BsmtFinSF1      0.0
BsmtFinType2    Unf
BsmtFinSF2      0.0
BsmtUnfSF       0.0
TotalBsmtSF     0.0
BsmtFullBath    0.0
BsmtHalfBath    0.0
Name: 2121, dtype: object

In [1174]:
Bsmt_df.groupby(['TotalBsmtSF'])['BsmtQual'].describe()

count unique  top freq
TotalBsmtSF                       
0.0             0      0  NaN  NaN
105.0           1      1   TA    1
160.0           1      1   Gd    1
173.0           0      0  NaN  NaN
190.0           1      1   Fa    1
...           ...    ...  ...  ...
3138.0          1      1   Ex    1
3200.0          1      1   Gd    1
3206.0          1      1   Gd    1
5095.0          1      1   Ex    1
6110.0          1      1   Ex    1

[1058 rows x 4 columns]

#### TotalBsmtSF 지하 면적의 총 평방 피트  
: Total square feet of basement area


In [1175]:
df['TotalBsmtSF'].isnull().sum()

0

TotalBsmtSF = 0인 경우에 BsmtQual의 결측치는 NA(지하실이 없다)로 처리

In [1177]:
df[df['TotalBsmtSF'] == 0]['BsmtQual'].head()

Id
18     NaN
40     NaN
91     NaN
103    NaN
157    NaN
Name: BsmtQual, dtype: object

- NA 대신 NB로 결측치 처리

In [1178]:
df['BsmtQual'][df['TotalBsmtSF']==0] = 'NB'

In [1179]:
df['BsmtQual'].isnull().sum()

2

In [1180]:
df[Bsmt][df['BsmtQual'].isnull()]

BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
Id                                                                          
2218      NaN       Fa           No          Unf         0.0          Unf   
2219      NaN       TA           No          Unf         0.0          Unf   

      BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  BsmtFullBath  BsmtHalfBath  
Id                                                                    
2218         0.0      173.0        173.0           0.0           0.0  
2219         0.0      356.0        356.0           0.0           0.0

In [1181]:
df.groupby('BsmtCond')['BsmtQual'].describe()

count unique top  freq
BsmtCond                       
Fa         103      3  TA    72
Gd         122      4  Gd    66
Po           5      3  Fa     3
TA        2606      5  TA  1177

In [1182]:
BsmtQual_mode =df['BsmtQual'].mode().iloc[0]
BsmtQual_mode

'TA'

In [1183]:
df['BsmtQual'].fillna(BsmtQual_mode, inplace=True)

In [1184]:
df['BsmtQual'].isnull().sum()

0

### BsmtCond 지하실의 전반적인 상태 평가
: Evaluates the general condition of the basement

       Ex	Excellent 우수
       Gd	Good 굿
       TA	Typical - slight dampness allowed 표준 - 약간의 습기 허용
       Fa	Fair - dampness or some cracking or settling 습기 또는 일부 균열 또는 침전
       Po	Poor - Severe cracking, settling, or wetness 심한 균열, 침하 또는 젖음
       NA	No Basement 지하실 없음

In [1185]:
df['BsmtCond'].isnull().sum()

81

In [1186]:
Bsmt_df = df[Bsmt]
Bsmt_df.isnull().sum()

BsmtQual         0
BsmtCond        81
BsmtExposure    81
BsmtFinType1    78
BsmtFinSF1       0
BsmtFinType2    79
BsmtFinSF2       0
BsmtUnfSF        0
TotalBsmtSF      0
BsmtFullBath     1
BsmtHalfBath     1
dtype: int64

In [1187]:
Bsmt_df[Bsmt_df['BsmtCond'].isnull()]

BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
Id                                                                          
18         NB      NaN          NaN          NaN         0.0          NaN   
40         NB      NaN          NaN          NaN         0.0          NaN   
91         NB      NaN          NaN          NaN         0.0          NaN   
103        NB      NaN          NaN          NaN         0.0          NaN   
157        NB      NaN          NaN          NaN         0.0          NaN   
...       ...      ...          ...          ...         ...          ...   
2804       NB      NaN          NaN          NaN         0.0          NaN   
2805       NB      NaN          NaN          NaN         0.0          NaN   
2825       NB      NaN          NaN          NaN         0.0          NaN   
2892       NB      NaN          NaN          NaN         0.0          NaN   
2905       NB      NaN          NaN          NaN         0.0          NaN   

      BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  BsmtFullBath  BsmtHalfBath  
Id                                                                    
18           0.0        0.0          0.0           0.0           0.0  
40           0.0        0.0          0.0           0.0           0.0  
91           0.0        0.0          0.0           0.0           0.0  
103          0.0        0.0          0.0           0.0           0.0  
157          0.0        0.0          0.0           0.0           0.0  
...          ...        ...          ...           ...           ...  
2804         0.0        0.0          0.0           0.0           0.0  
2805         0.0        0.0          0.0           0.0           0.0  
2825         0.0        0.0          0.0           0.0           0.0  
2892         0.0        0.0          0.0           0.0           0.0  
2905         0.0        0.0          0.0           0.0           0.0  

[81 rows x 11 columns]

- TotalBsmtSF = 0인 BsmtCond는 NB로 결측치 처리

In [1188]:
df['BsmtCond'][df['TotalBsmtSF']==0] = 'NB'

In [1189]:
df['BsmtCond'].isnull().sum()

3

In [1190]:
df[['BsmtCond','Electrical']][df['BsmtCond'].isnull()]

BsmtCond Electrical
Id                      
2041      NaN      SBrkr
2186      NaN      SBrkr
2525      NaN      SBrkr

In [1191]:
df.groupby('Electrical')['BsmtCond'].describe()

count unique top  freq
Electrical                       
FuseA        188      5  TA   152
FuseF         50      3  TA    33
FuseP          8      3  TA     6
Mix            1      1  Po     1
SBrkr       2668      5  TA  2414

- TotalBsmtSF가 0이 아닌 BsmtCond 남은 결측치는 최빈값으로 결측치 처리

In [1192]:
BsmtCond_mode = df['BsmtCond'].mode().iloc[0]
BsmtCond_mode

'TA'

In [1193]:
df['BsmtCond'].fillna(BsmtCond_mode, inplace=True)

In [1194]:
df['BsmtCond'].isnull().sum()

0

### BsmtExposure
: Refers to walkout or garden level walls 워크아웃 또는 정원 수준의 벽

       Gd	Good Exposure 노출 양호
       Av	Average Exposure (split levels or foyers typically score average or above) 노출 평균
       Mn	Mimimum Exposure 노출 최소
       No	No Exposure 노출 금지
       NA	No Basement 지하실 없음

In [1195]:
df[Bsmt].isnull().sum()

BsmtQual         0
BsmtCond         0
BsmtExposure    81
BsmtFinType1    78
BsmtFinSF1       0
BsmtFinType2    79
BsmtFinSF2       0
BsmtUnfSF        0
TotalBsmtSF      0
BsmtFullBath     1
BsmtHalfBath     1
dtype: int64

In [1197]:
Bsmt_df[df['BsmtExposure'].isnull()].head()

BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
Id                                                                         
18        NB      NaN          NaN          NaN         0.0          NaN   
40        NB      NaN          NaN          NaN         0.0          NaN   
91        NB      NaN          NaN          NaN         0.0          NaN   
103       NB      NaN          NaN          NaN         0.0          NaN   
157       NB      NaN          NaN          NaN         0.0          NaN   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  BsmtFullBath  BsmtHalfBath  
Id                                                                   
18          0.0        0.0          0.0           0.0           0.0  
40          0.0        0.0          0.0           0.0           0.0  
91          0.0        0.0          0.0           0.0           0.0  
103         0.0        0.0          0.0           0.0           0.0  
157         0.0        0.0          0.0           0.0           0.0

- 위와 동일하게 TotalBsmtSF = 0인 BsmtExposure 결측치는 NB로 처리

In [1198]:
df['BsmtExposure'][df['TotalBsmtSF']==0] = 'NB'

In [1199]:
df['BsmtExposure'].isnull().sum()

3

In [1200]:
df[Bsmt][df['BsmtExposure'].isnull()]

BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
Id                                                                          
949        Gd       TA          NaN          Unf         0.0          Unf   
1488       Gd       TA          NaN          Unf         0.0          Unf   
2349       Gd       TA          NaN          Unf         0.0          Unf   

      BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  BsmtFullBath  BsmtHalfBath  
Id                                                                    
949          0.0      936.0        936.0           0.0           0.0  
1488         0.0     1595.0       1595.0           0.0           0.0  
2349         0.0      725.0        725.0           0.0           0.0

In [1203]:
df.groupby(['BsmtFullBath','BsmtHalfBath'])['BsmtExposure'].describe().top

BsmtFullBath  BsmtHalfBath
0.0           0.0             No
              1.0             No
              2.0             Gd
1.0           0.0             No
              1.0             Gd
              2.0             Gd
2.0           0.0             Gd
3.0           0.0             Gd
Name: top, dtype: object

- BsmtExposure 남은 결측치는 No로 결측치 처리 

In [1204]:
df['BsmtExposure'].fillna('No', inplace=True)

In [1205]:
df['BsmtExposure'].isnull().sum()

0

### BsmtFinType1 완성된 지하실 면적 등급
: Rating of basement finished area

       GLQ	Good Living Quarters 좋은 생활관
       ALQ	Average Living Quarters 평균 거주지
       BLQ	Below Average Living Quarters	평균 생활권 이하
       Rec	Average Rec Room 평균 기록실
       LwQ	Low Quality 저품질
       Unf	Unfinshed 미완료 상태
       NA	No Basement 지하실 없음

In [1206]:
df[Bsmt][df['BsmtFinType1'].isnull()].head()

BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
Id                                                                         
18        NB       NB           NB          NaN         0.0          NaN   
40        NB       NB           NB          NaN         0.0          NaN   
91        NB       NB           NB          NaN         0.0          NaN   
103       NB       NB           NB          NaN         0.0          NaN   
157       NB       NB           NB          NaN         0.0          NaN   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  BsmtFullBath  BsmtHalfBath  
Id                                                                   
18          0.0        0.0          0.0           0.0           0.0  
40          0.0        0.0          0.0           0.0           0.0  
91          0.0        0.0          0.0           0.0           0.0  
103         0.0        0.0          0.0           0.0           0.0  
157         0.0        0.0          0.0           0.0           0.0

- TotalBsmtSF = 0인 BsmtFinType1 결측치는 NB로 처리

In [1207]:
df['BsmtFinType1'][df['TotalBsmtSF']==0] = 'NB'

In [1208]:
df['BsmtFinType1'].isnull().sum()

0

### BsmtFinType2 복수 유형인 경우 완성된 지하실 등급
: Rating of basement finished area (if multiple types)

       GLQ	Good Living Quarters 좋은 생활관
       ALQ	Average Living Quarters 평균 거주지
       BLQ	Below Average Living Quarters	평균 생활권 이하
       Rec	Average Rec Room 평균 기록실
       LwQ	Low Quality 저품질 
       Unf	Unfinshed 미완료 상태
       NA	No Basement 지하실 없음

In [1211]:
df[Bsmt][df['BsmtFinType2'].isnull()].head()

BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
Id                                                                         
18        NB       NB           NB           NB         0.0          NaN   
40        NB       NB           NB           NB         0.0          NaN   
91        NB       NB           NB           NB         0.0          NaN   
103       NB       NB           NB           NB         0.0          NaN   
157       NB       NB           NB           NB         0.0          NaN   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  BsmtFullBath  BsmtHalfBath  
Id                                                                   
18          0.0        0.0          0.0           0.0           0.0  
40          0.0        0.0          0.0           0.0           0.0  
91          0.0        0.0          0.0           0.0           0.0  
103         0.0        0.0          0.0           0.0           0.0  
157         0.0        0.0          0.0           0.0           0.0

- TotalBsmtSF가 0인 BsmtFinType2결측치는 NB(지하실 없음)로 처리

In [1212]:
df['BsmtFinType2'][df['TotalBsmtSF']==0] = 'NB'

In [1213]:
df['BsmtFinType2'].isnull().sum()

1

In [1214]:
df[Bsmt][df['BsmtFinType2'].isnull()]

BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
Id                                                                         
333       Gd       TA           No          GLQ      1124.0          NaN   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  BsmtFullBath  BsmtHalfBath  
Id                                                                   
333       479.0     1603.0       3206.0           1.0           0.0

- BsmtFinSF2 > 0 : 완성된 것을 의미하므로 BsmtFinType2이 Unf(미완성된)이 되면 안된다

In [1215]:
BsmtFinSF2_over_0 = df[Bsmt][df['BsmtFinSF2']>0]
BsmtFinSF2_over_0.head()

BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
Id                                                                        
8        Gd       TA           Mn          ALQ       859.0          BLQ   
25       TA       TA           Mn          Rec       188.0          ALQ   
27       TA       TA           Mn          BLQ       234.0          Rec   
43       Gd       TA           Av          ALQ       747.0          LwQ   
44       Gd       TA           Av          LwQ       280.0          BLQ   

    BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  BsmtFullBath  BsmtHalfBath  
Id                                                                  
8         32.0      216.0       1107.0           1.0           0.0  
25       668.0      204.0       1060.0           1.0           0.0  
27       486.0      180.0        900.0           0.0           1.0  
43        93.0        0.0        840.0           1.0           0.0  
44       491.0      167.0        938.0           1.0           0.0

- BsmtFinSF2 > 0 중에서 BsmtFinType2의 최빈값으로 결측치 처리

In [1216]:
BsmtFinType2_mode = BsmtFinSF2_over_0['BsmtFinType2'].mode().iloc[0]
BsmtFinType2_mode

'Rec'

In [1217]:
df['BsmtFinType2'].fillna(BsmtFinType2_mode, inplace=True)

In [1218]:
df['BsmtFinType2'].isnull().sum()

0

### BsmtFullBath 
: Basement full bathrooms

### BsmtHalfBath
: Basement half bathrooms

In [1230]:
df[Bsmt].isnull().sum()

BsmtQual        0
BsmtCond        0
BsmtExposure    0
BsmtFinType1    0
BsmtFinSF1      0
BsmtFinType2    0
BsmtFinSF2      0
BsmtUnfSF       0
TotalBsmtSF     0
BsmtFullBath    1
BsmtHalfBath    1
dtype: int64

In [1231]:
df[df['BsmtFullBath'].isnull()][Bsmt]

BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  \
Id                                                                          
2189       NB       NB           NB           NB         0.0           NB   

      BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  BsmtFullBath  BsmtHalfBath  
Id                                                                    
2189         0.0        0.0          0.0           NaN           NaN

- BsmtFullBath, BsmtHalfBath 0으로 결측치 채우기

In [1235]:
df['BsmtFullBath'].fillna(0, inplace=True)

In [1236]:
df['BsmtHalfBath'].fillna(0, inplace=True)

In [1237]:
df[Bsmt].isnull().sum()

BsmtQual        0
BsmtCond        0
BsmtExposure    0
BsmtFinType1    0
BsmtFinSF1      0
BsmtFinType2    0
BsmtFinSF2      0
BsmtUnfSF       0
TotalBsmtSF     0
BsmtFullBath    0
BsmtHalfBath    0
dtype: int64

### Electrical
: Electrical system

       SBrkr	Standard Circuit Breakers & Romex
       FuseA	Fuse Box over 60 AMP and all Romex wiring (Average)	
       FuseF	60 AMP Fuse Box and mostly Romex wiring (Fair)
       FuseP	60 AMP Fuse Box and mostly knob & tube wiring (poor)
       Mix	Mixed

In [1238]:
df['Electrical'].isnull().sum()

1

- 이전에 BsmtCond와 상관관계가 있었으므로 이를 이용하여 결측치 처리

In [1241]:
df.groupby('BsmtCond')['Electrical'].describe()

count unique    top  freq
BsmtCond                          
Fa         104      4  SBrkr    78
Gd         122      2  SBrkr   118
NB          79      4  SBrkr    55
Po           5      3  SBrkr     3
TA        2608      4  SBrkr  2417

In [1242]:
df['Electrical'].fillna('SBrkr', inplace=True)

In [1243]:
df['Electrical'].isnull().sum()

0

### KitchenQual
: Kitchen quality

       Ex	Excellent
       Gd	Good
       TA	Typical/Average
       Fa	Fair
       Po	Poor

In [1244]:
df['KitchenQual'].isnull().sum()

1

In [1247]:
df.groupby('KitchenAbvGr')['KitchenQual'].describe()

count unique top  freq
KitchenAbvGr                       
0                3      1  TA     3
1             2784      4  TA  1372
2              129      3  TA   115
3                2      1  TA     2

- TA로 결측치 처리

In [1248]:
df['KitchenQual'].fillna('TA', inplace=True)

In [1249]:
df['KitchenQual'].isnull().sum()

0

### Functional 홈 기능
: Home functionality (Assume typical unless deductions are warranted)

       Typ	Typical Functionality
       Min1	Minor Deductions 1
       Min2	Minor Deductions 2
       Mod	Moderate Deductions
       Maj1	Major Deductions 1
       Maj2	Major Deductions 2
       Sev	Severely Damaged
       Sal	Salvage only

In [1250]:
df['Functional'].isnull().sum()

2

In [1254]:
df['Functional'].value_counts()

Typ     2717
Min2      70
Min1      65
Mod       35
Maj1      19
Maj2       9
Sev        2
Name: Functional, dtype: int64

In [1255]:
df.groupby('Electrical')['Functional'].describe()

count unique   top  freq
Electrical                         
FuseA        187      6   Typ   162
FuseF         50      4   Typ    43
FuseP          8      1   Typ     8
Mix            1      1  Maj2     1
SBrkr       2671      7   Typ  2504

In [1256]:
df[['Electrical','Functional']][df['Functional'].isnull()]

Electrical Functional
Id                        
2217      FuseA        NaN
2474      SBrkr        NaN

In [1257]:
df['Functional'].fillna('Typ', inplace=True)

In [1258]:
df['Functional'].isnull().sum()

0

### FireplaceQu
: Fireplace quality

       Ex	Excellent - Exceptional Masonry Fireplace
       Gd	Good - Masonry Fireplace in main level
       TA	Average - Prefabricated Fireplace in main living area or Masonry Fireplace in basement
       Fa	Fair - Prefabricated Fireplace in basement
       Po	Poor - Ben Franklin Stove
       NA	No Fireplace

In [1260]:
df['FireplaceQu'].isnull().sum()

1420

In [1259]:
df['FireplaceQu'].value_counts()

Gd    744
TA    592
Fa     74
Po     46
Ex     43
Name: FireplaceQu, dtype: int64

In [1262]:
df.groupby('Fireplaces')['FireplaceQu'].describe()

count unique  top freq
Fireplaces                       
0              0      0  NaN  NaN
1           1268      5   Gd  627
2            219      4   Gd  112
3             11      4   Gd    5
4              1      1   TA    1

In [1264]:
df['FireplaceQu'][df['Fireplaces']==0] = 'NA'

In [1265]:
df['FireplaceQu'].isnull().sum()

0

### Garage

In [1266]:
Garage = []
for i in range(len(df_isnull_columns)):
    if 'Garage' in df_isnull_columns[i]:
        Garage.append(df_isnull_columns[i])
    else:
        pass
Garage

['GarageType',
 'GarageYrBlt',
 'GarageFinish',
 'GarageCars',
 'GarageArea',
 'GarageQual',
 'GarageCond']

In [1268]:
Garage_df = df[Garage]
Garage_df.head()

GarageType  GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual  \
Id                                                                           
1      Attchd       2003.0          RFn         2.0       548.0         TA   
2      Attchd       1976.0          RFn         2.0       460.0         TA   
3      Attchd       2001.0          RFn         2.0       608.0         TA   
4      Detchd       1998.0          Unf         3.0       642.0         TA   
5      Attchd       2000.0          RFn         3.0       836.0         TA   

   GarageCond  
Id             
1          TA  
2          TA  
3          TA  
4          TA  
5          TA

In [1269]:
Garage_df.isnull().sum()

GarageType      157
GarageYrBlt     159
GarageFinish    159
GarageCars        1
GarageArea        1
GarageQual      159
GarageCond      159
dtype: int64

### GarageType
: Garage location
		
       2Types	More than one type of garage
       Attchd	Attached to home
       Basment	Basement Garage
       BuiltIn	Built-In (Garage part of house - typically has room above garage)
       CarPort	Car Port
       Detchd	Detached from home
       NA	No Garage

In [1271]:
Garage_df[Garage_df['GarageType'].isnull()]

GarageType  GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual  \
Id                                                                             
40          NaN          NaN          NaN         0.0         0.0        NaN   
49          NaN          NaN          NaN         0.0         0.0        NaN   
79          NaN          NaN          NaN         0.0         0.0        NaN   
89          NaN          NaN          NaN         0.0         0.0        NaN   
90          NaN          NaN          NaN         0.0         0.0        NaN   
...         ...          ...          ...         ...         ...        ...   
2894        NaN          NaN          NaN         0.0         0.0        NaN   
2910        NaN          NaN          NaN         0.0         0.0        NaN   
2914        NaN          NaN          NaN         0.0         0.0        NaN   
2915        NaN          NaN          NaN         0.0         0.0        NaN   
2918        NaN          NaN          NaN         0.0         0.0        NaN   

     GarageCond  
Id               
40          NaN  
49          NaN  
79          NaN  
89          NaN  
90          NaN  
...         ...  
2894        NaN  
2910        NaN  
2914        NaN  
2915        NaN  
2918        NaN  

[157 rows x 7 columns]

- GarageArea = 0 => 차고지는 없다. => NA로 결측치 처리

In [1272]:
df['GarageType'][df['GarageArea']==0] = 'NA'

In [1273]:
df['GarageType'].isnull().sum()

0

### GarageYrBlt
: Year garage was built

In [1274]:
df['GarageYrBlt'][df['GarageArea']==0] = 'NA'

In [1275]:
df['GarageYrBlt'].isnull().sum()

2

In [1277]:
df[Garage][df['GarageYrBlt'].isnull()]

GarageType GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual  \
Id                                                                            
2127     Detchd         NaN          NaN         1.0       360.0        NaN   
2577     Detchd         NaN          NaN         NaN         NaN        NaN   

     GarageCond  
Id               
2127        NaN  
2577        NaN

In [1295]:
df['GarageYrBlt']

Id
1       2003.0
2       1976.0
3       2001.0
4       1998.0
5       2000.0
         ...  
2915        NA
2916    1970.0
2917    1960.0
2918        NA
2919    1993.0
Name: GarageYrBlt, Length: 2919, dtype: object

In [1287]:
df.groupby('GarageType')['GarageYrBlt'].describe().top

GarageType
2Types     1963.0
Attchd     2005.0
Basment    2005.0
BuiltIn    2003.0
CarPort    1920.0
Detchd     1950.0
NA             NA
Name: top, dtype: object

In [1296]:
df['GarageYrBlt'].fillna('1950.0', inplace=True)

In [1297]:
df['GarageYrBlt'].isnull().sum()

0

### GarageFinish
: Interior finish of the garage

       Fin	Finished
       RFn	Rough Finished	
       Unf	Unfinished
       NA	No Garage

In [1298]:
df['GarageFinish'].isnull().sum()

159

In [1301]:
df[Garage][df['GarageFinish'].isnull()].head()

GarageType GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual  \
Id                                                                          
40         NA          NA          NaN         0.0         0.0        NaN   
49         NA          NA          NaN         0.0         0.0        NaN   
79         NA          NA          NaN         0.0         0.0        NaN   
89         NA          NA          NaN         0.0         0.0        NaN   
90         NA          NA          NaN         0.0         0.0        NaN   

   GarageCond  
Id             
40        NaN  
49        NaN  
79        NaN  
89        NaN  
90        NaN

- GarageArea = 0 => NA로 결측치 처리

In [1304]:
df['GarageFinish'][df['GarageArea']==0] = 'NA'

In [1305]:
df['GarageFinish'].isnull().sum()

2

In [1306]:
df[Garage][df['GarageFinish'].isnull()]

GarageType GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual  \
Id                                                                            
2127     Detchd      1950.0          NaN         1.0       360.0        NaN   
2577     Detchd      1950.0          NaN         NaN         NaN        NaN   

     GarageCond  
Id               
2127        NaN  
2577        NaN

In [1309]:
df.groupby('GarageType')['GarageFinish'].describe()

count unique  top freq
GarageType                       
2Types        23      3  Unf   15
Attchd      1723      3  RFn  716
Basment       36      3  Unf   17
BuiltIn      186      3  Fin  127
CarPort       15      1  Unf   15
Detchd       777      3  Unf  719
NA           157      1   NA  157

In [1310]:
df['GarageFinish'].fillna('Unf', inplace=True)

In [1311]:
df['GarageFinish'].isnull().sum()

0

### GarageCars
: Size of garage in car capacity

In [1312]:
df['GarageCars'].isnull().sum()

1

In [1313]:
df[Garage][df['GarageCars'].isnull()]

GarageType GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual  \
Id                                                                            
2577     Detchd      1950.0          Unf         NaN         NaN        NaN   

     GarageCond  
Id               
2577        NaN

- 최빈값으로 결측치 처리

In [1318]:
GarageCars_mode = df['GarageCars'].mode().iloc[0]
GarageCars_mode

2.0

In [1319]:
df['GarageCars'].fillna(GarageCars_mode, inplace=True)

In [1320]:
df['GarageCars'].isnull().sum()

0

### GarageArea
: Size of garage in square feet

In [1321]:
df['GarageArea'].isnull().sum()

1

In [1322]:
df[Garage][df['GarageArea'].isnull()]

GarageType GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual  \
Id                                                                            
2577     Detchd      1950.0          Unf         2.0         NaN        NaN   

     GarageCond  
Id               
2577        NaN

- 최빈값으로 결측치 처리

In [1323]:
df.groupby('GarageCars')['GarageArea'].describe()

count         mean         std     min     25%     50%      75%  \
GarageCars                                                                     
0.0          157.0     0.000000    0.000000     0.0     0.0     0.0     0.00   
1.0          776.0   300.210052   82.169791   100.0   252.0   288.0   328.50   
2.0         1594.0   519.432873   95.637581   264.0   452.0   506.0   576.00   
3.0          374.0   808.328877  130.957072   350.0   721.0   807.5   873.50   
4.0           16.0   963.125000  258.157284   480.0   811.0   892.0  1068.25   
5.0            1.0  1184.000000         NaN  1184.0  1184.0  1184.0  1184.00   

               max  
GarageCars          
0.0            0.0  
1.0          924.0  
2.0         1418.0  
3.0         1390.0  
4.0         1488.0  
5.0         1184.0

In [1331]:
GarageArea_round_mean = round(df.groupby('GarageCars')['GarageArea'].transform('mean'),0)
GarageArea_round_mean

Id
1       519.0
2       519.0
3       519.0
4       808.0
5       808.0
        ...  
2915      0.0
2916    300.0
2917    519.0
2918      0.0
2919    808.0
Name: GarageArea, Length: 2919, dtype: float64

In [1332]:
df['GarageArea'].fillna(GarageArea_round_mean, inplace=True)

In [1333]:
df['GarageArea'].isnull().sum()

0

In [1334]:
df.iloc[2576]['GarageArea']

519.0

### GarageQual
: Garage quality

       Ex	Excellent
       Gd	Good
       TA	Typical/Average
       Fa	Fair
       Po	Poor
       NA	No Garage

In [1335]:
df['GarageQual'].isnull().sum()

159

In [1336]:
df[Garage][df['GarageQual'].isnull()]

GarageType GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual  \
Id                                                                            
40           NA          NA           NA         0.0         0.0        NaN   
49           NA          NA           NA         0.0         0.0        NaN   
79           NA          NA           NA         0.0         0.0        NaN   
89           NA          NA           NA         0.0         0.0        NaN   
90           NA          NA           NA         0.0         0.0        NaN   
...         ...         ...          ...         ...         ...        ...   
2894         NA          NA           NA         0.0         0.0        NaN   
2910         NA          NA           NA         0.0         0.0        NaN   
2914         NA          NA           NA         0.0         0.0        NaN   
2915         NA          NA           NA         0.0         0.0        NaN   
2918         NA          NA           NA         0.0         0.0        NaN   

     GarageCond  
Id               
40          NaN  
49          NaN  
79          NaN  
89          NaN  
90          NaN  
...         ...  
2894        NaN  
2910        NaN  
2914        NaN  
2915        NaN  
2918        NaN  

[159 rows x 7 columns]

In [1337]:
df['GarageQual'][df['GarageArea']==0] = 'NA'

In [1338]:
df['GarageQual'].isnull().sum()

2

In [1339]:
df[Garage][df['GarageQual'].isnull()]

GarageType GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual  \
Id                                                                            
2127     Detchd      1950.0          Unf         1.0       360.0        NaN   
2577     Detchd      1950.0          Unf         2.0       519.0        NaN   

     GarageCond  
Id               
2127        NaN  
2577        NaN

In [1346]:
GarageQual_mode = df['GarageQual'].mode().iloc[0]
GarageQual_mode

'TA'

In [1347]:
df['GarageQual'].fillna(GarageQual_mode, inplace=True)

In [1348]:
df['GarageQual'].isnull().sum()

0

### GarageCond
: Garage condition

       Ex	Excellent
       Gd	Good
       TA	Typical/Average
       Fa	Fair
       Po	Poor
       NA	No Garage
		

In [1349]:
df['GarageCond'].isnull().sum()

159

In [1350]:
df[Garage][df['GarageCond'].isnull()]

GarageType GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual  \
Id                                                                            
40           NA          NA           NA         0.0         0.0         NA   
49           NA          NA           NA         0.0         0.0         NA   
79           NA          NA           NA         0.0         0.0         NA   
89           NA          NA           NA         0.0         0.0         NA   
90           NA          NA           NA         0.0         0.0         NA   
...         ...         ...          ...         ...         ...        ...   
2894         NA          NA           NA         0.0         0.0         NA   
2910         NA          NA           NA         0.0         0.0         NA   
2914         NA          NA           NA         0.0         0.0         NA   
2915         NA          NA           NA         0.0         0.0         NA   
2918         NA          NA           NA         0.0         0.0         NA   

     GarageCond  
Id               
40          NaN  
49          NaN  
79          NaN  
89          NaN  
90          NaN  
...         ...  
2894        NaN  
2910        NaN  
2914        NaN  
2915        NaN  
2918        NaN  

[159 rows x 7 columns]

In [1351]:
df['GarageCond'][df['GarageArea'] == 0] = 'NA'

In [1352]:
df['GarageCond'].isnull().sum()

2

In [1353]:
df[Garage][df['GarageCond'].isnull()]

GarageType GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual  \
Id                                                                            
2127     Detchd      1950.0          Unf         1.0       360.0         TA   
2577     Detchd      1950.0          Unf         2.0       519.0         TA   

     GarageCond  
Id               
2127        NaN  
2577        NaN

In [1354]:
df.groupby('GarageQual')['GarageCond'].describe()

count unique top  freq
GarageQual                       
Ex             3      2  Ex     2
Fa           124      3  TA    77
Gd            24      3  TA    17
NA           157      1  NA   157
Po             5      2  Po     4
TA          2604      4  TA  2559

In [1355]:
df['GarageCond'].fillna('TA', inplace= True)

In [1356]:
df['GarageCond'].isnull().sum()

0

### SaleType
: Type of sale
		
       WD 	Warranty Deed - Conventional
       CWD	Warranty Deed - Cash
       VWD	Warranty Deed - VA Loan
       New	Home just constructed and sold
       COD	Court Officer Deed/Estate
       Con	Contract 15% Down payment regular terms
       ConLw	Contract Low Down payment and low interest
       ConLI	Contract Low Interest
       ConLD	Contract Low Down
       Oth	Other

In [1357]:
df['SaleType'].isnull().sum()

1

In [1360]:
df.groupby('SaleCondition')['SaleType'].describe()

count unique  top  freq
SaleCondition                        
Abnorml         190      6   WD   133
AdjLand          12      1   WD    12
Alloca           24      1   WD    24
Family           46      7   WD    38
Normal         2401      8   WD  2314
Partial         245      4  New   239

In [1361]:
df[['SaleCondition','SaleType']][df['SaleType'].isnull()]

SaleCondition SaleType
Id                         
2490        Normal      NaN

In [1362]:
df['SaleType'].fillna('WD', inplace=True)

In [1363]:
df['SaleType'].isnull().sum()

0

### 결측치 모두 처리 완료

In [1366]:
np.sum(df.isnull().sum())

0

### One-Hot Encoding

- 원본데이터 훼손하지 않기위해 Copy

In [1386]:
data = df.copy()
data.head()

MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
Id                                                                          
1           60       RL         65.0     8450   Pave      Reg         Lvl   
2           20       RL         80.0     9600   Pave      Reg         Lvl   
3           60       RL         68.0    11250   Pave      IR1         Lvl   
4           70       RL         60.0     9550   Pave      IR1         Lvl   
5           60       RL         84.0    14260   Pave      IR1         Lvl   

   Utilities LotConfig LandSlope  ... OpenPorchSF EnclosedPorch 3SsnPorch  \
Id                                ...                                       
1     AllPub    Inside       Gtl  ...          61             0         0   
2     AllPub       FR2       Gtl  ...           0             0         0   
3     AllPub    Inside       Gtl  ...          42             0         0   
4     AllPub    Corner       Gtl  ...          35           272         0   
5     AllPub       FR2       Gtl  ...          84             0         0   

   ScreenPorch PoolArea  MiscVal  MoSold  YrSold  SaleType SaleCondition  
Id                                                                        
1            0        0        0       2    2008        WD        Normal  
2            0        0        0       5    2007        WD        Normal  
3            0        0        0       9    2008        WD        Normal  
4            0        0        0       2    2006        WD       Abnorml  
5            0        0        0      12    2008        WD        Normal  

[5 rows x 75 columns]

In [1426]:
df_object_columns

array(['MSZoning', 'Street', 'LotShape', 'LandContour', 'Utilities',
       'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1',
       'Condition2', 'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl',
       'Exterior1st', 'Exterior2nd', 'MasVnrType', 'ExterQual',
       'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure',
       'BsmtFinType1', 'BsmtFinType2', 'Heating', 'HeatingQC',
       'CentralAir', 'Electrical', 'KitchenQual', 'Functional',
       'FireplaceQu', 'GarageType', 'GarageFinish', 'GarageQual',
       'GarageCond', 'PavedDrive', 'SaleType', 'SaleCondition'],
      dtype=object)

In [1447]:
object_data = data[object_columns]
object_data.head()

MSZoning Street LotShape LandContour Utilities LotConfig LandSlope  \
Id                                                                      
1        RL   Pave      Reg         Lvl    AllPub    Inside       Gtl   
2        RL   Pave      Reg         Lvl    AllPub       FR2       Gtl   
3        RL   Pave      IR1         Lvl    AllPub    Inside       Gtl   
4        RL   Pave      IR1         Lvl    AllPub    Corner       Gtl   
5        RL   Pave      IR1         Lvl    AllPub       FR2       Gtl   

   Neighborhood Condition1 Condition2  ... Functional FireplaceQu GarageType  \
Id                                     ...                                     
1       CollgCr       Norm       Norm  ...        Typ          NA     Attchd   
2       Veenker      Feedr       Norm  ...        Typ          TA     Attchd   
3       CollgCr       Norm       Norm  ...        Typ          TA     Attchd   
4       Crawfor       Norm       Norm  ...        Typ          Gd     Detchd   
5       NoRidge       Norm       Norm  ...        Typ          TA     Attchd   

   GarageYrBlt GarageFinish GarageQual GarageCond PavedDrive SaleType  \
Id                                                                      
1       2003.0          RFn         TA         TA          Y       WD   
2       1976.0          RFn         TA         TA          Y       WD   
3       2001.0          RFn         TA         TA          Y       WD   
4       1998.0          Unf         TA         TA          Y       WD   
5       2000.0          RFn         TA         TA          Y       WD   

   SaleCondition  
Id                
1         Normal  
2         Normal  
3         Normal  
4        Abnorml  
5         Normal  

[5 rows x 40 columns]

In [1448]:
object_one_hot_data = pd.get_dummies(object_data)
object_one_hot_data.head()

MSZoning_C (all)  MSZoning_FV  MSZoning_RH  MSZoning_RL  MSZoning_RM  \
Id                                                                         
1                  0            0            0            1            0   
2                  0            0            0            1            0   
3                  0            0            0            1            0   
4                  0            0            0            1            0   
5                  0            0            0            1            0   

    Street_Grvl  Street_Pave  LotShape_IR1  LotShape_IR2  LotShape_IR3  ...  \
Id                                                                      ...   
1             0            1             0             0             0  ...   
2             0            1             0             0             0  ...   
3             0            1             1             0             0  ...   
4             0            1             1             0             0  ...   
5             0            1             1             0             0  ...   

    SaleType_ConLw  SaleType_New  SaleType_Oth  SaleType_WD  \
Id                                                            
1                0             0             0            1   
2                0             0             0            1   
3                0             0             0            1   
4                0             0             0            1   
5                0             0             0            1   

    SaleCondition_Abnorml  SaleCondition_AdjLand  SaleCondition_Alloca  \
Id                                                                       
1                       0                      0                     0   
2                       0                      0                     0   
3                       0                      0                     0   
4                       1                      0                     0   
5                       0                      0                     0   

    SaleCondition_Family  SaleCondition_Normal  SaleCondition_Partial  
Id                                                                     
1                      0                     1                      0  
2                      0                     1                      0  
3                      0                     1                      0  
4                      0                     0                      0  
5                      0                     1                      0  

[5 rows x 354 columns]

In [1445]:
numeric_data = data[numeric_columns]
numeric_data.head()

MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
Id                                                                          
1           60         65.0     8450            7            5       2003   
2           20         80.0     9600            6            8       1976   
3           60         68.0    11250            7            5       2001   
4           70         60.0     9550            7            5       1915   
5           60         84.0    14260            8            5       2000   

    YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  ...  GarageArea  \
Id                                                    ...               
1           2003       196.0       706.0         0.0  ...       548.0   
2           1976         0.0       978.0         0.0  ...       460.0   
3           2002       162.0       486.0         0.0  ...       608.0   
4           1970         0.0       216.0         0.0  ...       642.0   
5           2000       350.0       655.0         0.0  ...       836.0   

    WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea  \
Id                                                                             
1            0           61              0          0            0         0   
2          298            0              0          0            0         0   
3            0           42              0          0            0         0   
4            0           35            272          0            0         0   
5          192           84              0          0            0         0   

    MiscVal  MoSold  YrSold  
Id                           
1         0       2    2008  
2         0       5    2007  
3         0       9    2008  
4         0       2    2006  
5         0      12    2008  

[5 rows x 35 columns]

In [1446]:
numeric_data_normalization = (numeric_data - np.mean(numeric_data)) / np.std(numeric_data)
numeric_data_normalization.head()

MSSubClass  LotFrontage   LotArea  OverallQual  OverallCond  YearBuilt  \
Id                                                                           
1     0.067331    -0.201161 -0.217879     0.646183    -0.507284   1.046258   
2    -0.873616     0.490271 -0.072044    -0.063185     2.188279   0.154764   
3     0.067331    -0.062874  0.137197     0.646183    -0.507284   0.980221   
4     0.302568    -0.431638 -0.078385     0.646183    -0.507284  -1.859351   
5     0.067331     0.674653  0.518903     1.355551    -0.507284   0.947203   

    YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  ...  GarageArea  \
Id                                                    ...               
1       0.896833    0.522021    0.581145   -0.293025  ...    0.348824   
2      -0.395604   -0.573704    1.178255   -0.293025  ...   -0.059865   
3       0.848965    0.331946    0.098189   -0.293025  ...    0.627475   
4      -0.682812   -0.573704   -0.494529   -0.293025  ...    0.785378   
5       0.753229    1.382948    0.469187   -0.293025  ...    1.686351   

    WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea  \
Id                                                                             
1    -0.740760     0.200006      -0.359601  -0.103331    -0.285935  -0.06315   
2     1.614879    -0.702843      -0.359601  -0.103331    -0.285935  -0.06315   
3    -0.740760    -0.081209      -0.359601  -0.103331    -0.285935  -0.06315   
4    -0.740760    -0.184815       3.874967  -0.103331    -0.285935  -0.06315   
5     0.776967     0.540424      -0.359601  -0.103331    -0.285935  -0.06315   

     MiscVal    MoSold    YrSold  
Id                                
1  -0.089592 -1.552184  0.157646  
2  -0.089592 -0.446925 -0.602962  
3  -0.089592  1.026753  0.157646  
4  -0.089592 -1.552184 -1.363569  
5  -0.089592  2.132012  0.157646  

[5 rows x 35 columns]

In [1449]:
one_hot_df = pd.concat([numeric_data_normalization, object_one_hot_data], axis =1)
one_hot_df.head()

MSSubClass  LotFrontage   LotArea  OverallQual  OverallCond  YearBuilt  \
Id                                                                           
1     0.067331    -0.201161 -0.217879     0.646183    -0.507284   1.046258   
2    -0.873616     0.490271 -0.072044    -0.063185     2.188279   0.154764   
3     0.067331    -0.062874  0.137197     0.646183    -0.507284   0.980221   
4     0.302568    -0.431638 -0.078385     0.646183    -0.507284  -1.859351   
5     0.067331     0.674653  0.518903     1.355551    -0.507284   0.947203   

    YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  ...  SaleType_ConLw  \
Id                                                    ...                   
1       0.896833    0.522021    0.581145   -0.293025  ...               0   
2      -0.395604   -0.573704    1.178255   -0.293025  ...               0   
3       0.848965    0.331946    0.098189   -0.293025  ...               0   
4      -0.682812   -0.573704   -0.494529   -0.293025  ...               0   
5       0.753229    1.382948    0.469187   -0.293025  ...               0   

    SaleType_New  SaleType_Oth  SaleType_WD  SaleCondition_Abnorml  \
Id                                                                   
1              0             0            1                      0   
2              0             0            1                      0   
3              0             0            1                      0   
4              0             0            1                      1   
5              0             0            1                      0   

    SaleCondition_AdjLand  SaleCondition_Alloca  SaleCondition_Family  \
Id                                                                      
1                       0                     0                     0   
2                       0                     0                     0   
3                       0                     0                     0   
4                       0                     0                     0   
5                       0                     0                     0   

    SaleCondition_Normal  SaleCondition_Partial  
Id                                               
1                      1                      0  
2                      1                      0  
3                      1                      0  
4                      0                      0  
5                      1                      0  

[5 rows x 389 columns]

In [1450]:
number_of_train_df

1460

In [1451]:
X_train = one_hot_df[:number_of_train_df]
X_train.tail()

MSSubClass  LotFrontage   LotArea  OverallQual  OverallCond  YearBuilt  \
Id                                                                             
1456    0.067331    -0.339447 -0.285470    -0.063185    -0.507284   0.914184   
1457   -0.873616     0.720749  0.381311    -0.063185     0.391237   0.220801   
1458    0.302568    -0.155065 -0.142806     0.646183     3.086800  -1.000876   
1459   -0.873616    -0.062874 -0.057207    -0.772552     0.391237  -0.703711   
1460   -0.873616     0.259794 -0.029308    -0.772552     0.391237  -0.208437   

      YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  ...  SaleType_ConLw  \
Id                                                      ...                   
1456      0.753229   -0.573704   -0.968705   -0.293025  ...               0   
1457      0.178812    0.091557    0.765547    0.670616  ...               0   
1458      1.040437   -0.573704   -0.365009   -0.293025  ...               0   
1459      0.561757   -0.573704   -0.861137    5.790328  ...               0   
1460     -0.922153   -0.573704    0.853357    1.421428  ...               0   

      SaleType_New  SaleType_Oth  SaleType_WD  SaleCondition_Abnorml  \
Id                                                                     
1456             0             0            1                      0   
1457             0             0            1                      0   
1458             0             0            1                      0   
1459             0             0            1                      0   
1460             0             0            1                      0   

      SaleCondition_AdjLand  SaleCondition_Alloca  SaleCondition_Family  \
Id                                                                        
1456                      0                     0                     0   
1457                      0                     0                     0   
1458                      0                     0                     0   
1459                      0                     0                     0   
1460                      0                     0                     0   

      SaleCondition_Normal  SaleCondition_Partial  
Id                                                 
1456                     1                      0  
1457                     1                      0  
1458                     1                      0  
1459                     1                      0  
1460                     1                      0  

[5 rows x 389 columns]

In [1452]:
X_test = one_hot_df[number_of_train_df:]
X_test.head()

MSSubClass  LotFrontage   LotArea  OverallQual  OverallCond  YearBuilt  \
Id                                                                             
1461   -0.873616     0.490271  0.184371    -0.772552     0.391237  -0.340510   
1462   -0.873616     0.536367  0.519791    -0.063185     0.391237  -0.439565   
1463    0.067331     0.213699  0.464374    -0.772552    -0.507284   0.848148   
1464    0.067331     0.398080 -0.024109    -0.063185     0.391237   0.881166   
1465    1.478753    -1.215261 -0.654748     1.355551    -0.507284   0.683057   

      YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  ...  SaleType_ConLw  \
Id                                                      ...                   
1461     -1.113625   -0.573704    0.058675    0.558290  ...               0   
1462     -1.257229    0.030063    1.057516   -0.293025  ...               0   
1463      0.657493   -0.573704    0.767742   -0.293025  ...               0   
1464      0.657493   -0.461896    0.352839   -0.293025  ...               0   
1465      0.370284   -0.573704   -0.391353   -0.293025  ...               0   

      SaleType_New  SaleType_Oth  SaleType_WD  SaleCondition_Abnorml  \
Id                                                                     
1461             0             0            1                      0   
1462             0             0            1                      0   
1463             0             0            1                      0   
1464             0             0            1                      0   
1465             0             0            1                      0   

      SaleCondition_AdjLand  SaleCondition_Alloca  SaleCondition_Family  \
Id                                                                        
1461                      0                     0                     0   
1462                      0                     0                     0   
1463                      0                     0                     0   
1464                      0                     0                     0   
1465                      0                     0                     0   

      SaleCondition_Normal  SaleCondition_Partial  
Id                                                 
1461                     1                      0  
1462                     1                      0  
1463                     1                      0  
1464                     1                      0  
1465                     1                      0  

[5 rows x 389 columns]

In [1453]:
y_true

Id
1       208500.0
2       181500.0
3       223500.0
4       140000.0
5       250000.0
          ...   
1456    175000.0
1457    210000.0
1458    266500.0
1459    142125.0
1460    147500.0
Name: SalePrice, Length: 1460, dtype: float64

### 예측

In [1455]:
from sklearn.model_selection import KFold
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import GridSearchCV

In [1456]:
param_grid = {
    'n_estimators' : [50,150,250,300],
    'max_features' : ['auto','sqrt','log2'],
    'max_depth' : [10,20,30,40]
}

In [1457]:
kfold = KFold(n_splits=5, shuffle=True)
reg = RandomForestRegressor()
grid = GridSearchCV(estimator=reg, param_grid=param_grid, scoring='neg_mean_squared_error', cv =kfold, verbose = 2, n_jobs=-1)

In [1458]:
grid.fit(X_train,y_true)

Fitting 5 folds for each of 48 candidates, totalling 240 fits


GridSearchCV(cv=KFold(n_splits=5, random_state=None, shuffle=True),
             estimator=RandomForestRegressor(), n_jobs=-1,
             param_grid={'max_depth': [10, 20, 30, 40],
                         'max_features': ['auto', 'sqrt', 'log2'],
                         'n_estimators': [50, 150, 250, 300]},
             scoring='neg_mean_squared_error', verbose=2)

In [1459]:
grid.best_score_, grid.best_params_

(-814518029.4848093,
 {'max_depth': 30, 'max_features': 'auto', 'n_estimators': 50})

In [1460]:
y_pred = grid.predict(X_test)
y_pred

array([127045.  , 156918.8 , 172320.8 , ..., 154269.92, 112971.  ,
       227483.18])

In [1465]:
submission_columns = ['Id', 'SalePrice']
idx = (df[number_of_train_df:].index).tolist()

submission_df = pd.DataFrame([idx, y_pred]).T
submission_df.columns = submission_columns
for col_name in submission_columns:
    if col_name == 'Id':
        submission_df[col_name] = submission_df[col_name].astype(int)
    else:
        submission_df[col_name] = submission_df[col_name].astype(float)
        
submission_df.to_csv("submission.csv", index = False)